### PR TITLE
Initialize GC state as soon as possible and update it on (re)connections

### PR DESCRIPTION
**Pull request by agarwal-navin**
Currently, the GC state from the base snapshot is initialized when the container transitions to write mode. This means that the unreferenced tracking of nodes do not begin until container is in write mode. There are a couple of implications of this:
1. Any loads / changes to inactive, tombstone or sweep ready objects before container transitions to write mode are not detected.
2. If a container receives a GC sweep op before it transitions to write mode, it will delete sweep ready nodes. However, when GC is initialized, it will start tracking these deleted nodes as unrefenced leading to inconsistent state.

This PR initializes GC state when the container runtime loads, if a current reference timestamp. Then every time there is a new connection, if GC state is initialized, it updates the unreferened state based on the current reference timestamp and if GC is not initialized, it initializes it.
This ensures that any changes or loads to unreferenced nodes are detected and that the GC sweep op will be processed after the GC state has been initialized.

Note: This PR also converts the `TombstoneReadyObject_*` and `SweepReadyObject_*` logs to generic instead of error. These are already logged separately so they are mostly redundant. However, there are certain scenarios where these are still useful such as in interactive clients, when an object is not tombstone-ready / sweep-ready during initialization but later becomes so.

[AB#5760](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/5760)

### CODE REVIEW

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts [hunk 1]
```diff
@@ -350,115 +358,59 @@ export class GarbageCollector implements IGarbageCollector {
...
-	 * @param currentReferenceTimestampMs - The current reference timestamp for marking unreferenced nodes' unreferenced
-	 * timestamp.
+	 * Initialize the GC state if not already initialized. If GC state is already initialized, update the unreferenced
+	 * state tracking as per the current reference timestamp.
 	 */
-	private updateStateFromSnapshotData(
-		snapshotData: IGarbageCollectionSnapshotData | undefined,
-		currentReferenceTimestampMs: number,
-	) {
-		/**
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/gc/garbageCollection.ts:367]: Note: The logic to reset the unreferenced node state, update tombstone and deleted nodes were not required anymore because of the removal of refresh from snapshot flow. The only thing needed is to initialize `this.unreferencedNodeState` and `this.gcDataFromLastRun`. That logic got moved into `this.initializeGCStateFromBaseSnapshotP`

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts [hunk 2]
```diff
@@ -350,115 +358,59 @@ export class GarbageCollector implements IGarbageCollector {
...
-	 * Update state from the given snapshot data. This is done during load and during refreshing state from a snapshot.
-	 * All current tracking is reset and updated from the data in the snapshot.
-	 * @param snapshotData - The snapshot data to update state from. If this is undefined, all GC state and tracking
-	 * is reset.
-	 * @param currentReferenceTimestampMs - The current reference timestamp for marking unreferenced nodes' unreferenced
-	 * timestamp.
+	 * Initialize the GC state if not already initialized. If GC state is already initialized, update the unreferenced
+	 * state tracking as per the current reference timestamp.
 	 */
-	private updateStateFromSnapshotData(
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/garbageCollection.ts:363]: Glad to see this gone!  Reading over the code, looks like it had some gaps anyway...

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts [hunk 3]
```diff
@@ -258,36 +255,47 @@ export class GarbageCollector implements IGarbageCollector {
 		 * Set up the initializer which initializes the GC state from the data in base snapshot. This is done when
 		 * connected in write mode or when GC runs the first time. It sets up all unreferenced nodes from the base
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/garbageCollection.ts:256]: Update comment

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts [hunk 4]
```diff
@@ -258,36 +255,47 @@ export class GarbageCollector implements IGarbageCollector {
...
 		 * GC state and updates their inactive or sweep-ready state.
+		 *
+		 * Must only be called if there is a current reference timestamp.
 		 */
 		this.initializeGCStateFromBaseSnapshotP = new LazyPromise<void>(async () => {
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/garbageCollection.ts:261]: I'm trying to think if this makes more sense to turn into a regular function that takes in `currentReferenceTimestamp`.  I'll play around with it once I finish the review.

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts [hunk 5]
```diff
@@ -258,36 +255,47 @@ export class GarbageCollector implements IGarbageCollector {
...
+
 			/**
 			 * The base snapshot data will not be present if the container is loaded from:
 			 * 1. The first summary created by the detached container.
 			 * 2. A summary that was generated with GC disabled.
 			 * 3. A summary that was generated before GC even existed.
 			 */
 			const baseSnapshotData = await this.baseSnapshotDataP;
-			if (baseSnapshotData === undefined) {
+			if (baseSnapshotData?.gcState === undefined) {
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/garbageCollection.ts:275]: You should add a 4th case in the comment above where the GC Version doesn't match, now that this early return keys off of `gcState` in particular.

Comment (markfields) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts:275]: Oh, actually looks like this may be a bug/mistake.  In this case, we'll now skip the `this.summaryStateTracker.initializeBaseState` call below, but we should still be doing that if `gcState` is undefined but `baseSnapshotData` is not.

Comment (markfields) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts:275]: Also do you need to do `this.gcDataFromLastRun = undefined` in this case?  Or not since this is only for initialization and not refresh.

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts [hunk 6]
```diff
@@ -350,115 +358,59 @@ export class GarbageCollector implements IGarbageCollector {
 			this.tombstones = Array.from(baseSnapshotData.tombstones);
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/garbageCollection.ts:358]: Nit: The function comment seems out of date, at first it sounds like it's only about Tombstones.

Comment (markfields) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts:358]: More importantly it should be updated to explain that it's doing full init including the core GC Data, not just Tombstoned/Deleted nodes.

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts [hunk 7]
```diff
@@ -350,115 +358,59 @@ export class GarbageCollector implements IGarbageCollector {
...
-					// Call container runtime to delete these nodes and add deleted nodes to this.deletedNodes.
-				}
-			}
-		} else if (this.configs.tombstoneMode) {
-			// The snapshot may contain more or fewer tombstone nodes than this client. Update tombstone state and
-			// notify the runtime to update its state as well.
-			this.tombstones = snapshotData?.tombstones ? Array.from(snapshotData.tombstones) : [];
-			this.runtime.updateTombstonedRoutes(this.tombstones);
+	private async initializeOrUpdateGCState() {
+		const currentReferenceTimestampMs = this.runtime.getCurrentReferenceTimestampMs();
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/garbageCollection.ts:370]: I don't see an update to this function's logic to pull the timestamp from the IConnect message or whatever - is that included in this PR?  So this means we'll miss out on proper tracking of old stuff until we're in write mode, I think?

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts [hunk 8]
```diff
@@ -350,115 +358,59 @@ export class GarbageCollector implements IGarbageCollector {
...
-		 * sweep in phases and we want to track when inactive and sweep-ready objects are used in any client.
+		 * During GC initialization and during connections in read mode, it is possible that either no ops are
+		 * processed or only trailing ops are processed. This means that the GC state is not initialized or initialized
+		 * with an older reference timestamp. So, doing this on every connection will keep the unreferenced state
+		 * tracking up-to-date.
 		 */
-		if (this.activeConnection() && this.configs.shouldRunGC) {
-			this.initializeGCStateFromBaseSnapshotP.catch((error) => {});
+		if (connected && this.configs.shouldRunGC) {
+			this.initializeOrUpdateGCState().catch((error) => {
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/garbageCollection.ts:405]: Could this happen _before_ all trailing ops have processed?  In the case where the `currentReferenceTimestampMs` is based on the `IConnect` message, I think that could cause some trouble.  Seems like that's not part of this PR, so it's safe, it's just using this to periodically reallign with the op stream that's been processed so far.  Does that sound right?

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts [hunk 9]
```diff
@@ -350,115 +358,59 @@ export class GarbageCollector implements IGarbageCollector {
...
+				this.mc.logger.sendErrorEvent(
+					{
+						eventName: "GCInitializationOrUpdateFailed",
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/garbageCollection.ts:408]: log `clientId`?  Or is it on this already from the logger?

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:1]: Feels like this file should have some test coverage that would update with this change. But I can also imagine why it might not be the case.

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts [hunk 10]
```diff
@@ -173,10 +173,6 @@ describeCompat("GC data store sweep tests", "NoCompat", (getTestObjectProvider)
 		itExpects(
 			"Send ops fails for swept datastores in summarizing container loaded before sweep timeout",
 			[
-				{
-					eventName: "fluid:telemetry:Summarizer:Running:SweepReadyObject_Loaded",
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts:177]: I'm having a little bit of mixed feelings about these logs going away. They are a pain to maintain, they're always popping up in tests everywhere.  But I also like having test coverage _somewhere_ to prevent regressions, so when we see these logs we can be confident we're interpreting them properly.

Might not be worth it, thinking out loud.

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts [hunk 11]
```diff
@@ -315,7 +311,7 @@ describeCompat("GC attachment blob tombstone tests", "NoCompat", (getTestObjectP
 				container2.close();
 
 				// Reference the blob in the main container where it's not a tombstone yet. This should un-tombstone the
-				// blob. It will result in a SweepReadyObject_Revived error log.
+				// blob.
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts:314]: I'd leave this and just remove the word "error"

**Review by markfields — COMMENTED** [review]
Can you add some kind of test coverage for the change?  Even just one test, maybe in the trailing ops file, that covers a key case here.

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts (hunk 5 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts:275]: Oh good catch!

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts:275]: > Also do you need to do this.gcDataFromLastRun = undefined in this case? Or not since this is only for initialization and not refresh.

Yeah, this only happens on initialization so its not needed anymore.

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts (hunk 7 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts:370]: I decided not to use the `IConnected` or `IClient` timestamp because that code is would be littered with undefined checks. See the containerRuntime.ts in this file for how we used to it before - https://github.com/microsoft/FluidFramework/pull/9145/files.

We may lose some tracking for old stuff for a while, yes. In the worst case, until the container moves to write mode. However, in most cases, there would be ops for a doc and at least some client would transition to write mode and will get updated tracking. The worst case is when a document is opened in read-mode without any activity. I don't think its worth optimizing for that case. What do you think?

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts (hunk 4 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts:261]: FYI now that there's only one callsite (after you remove the one in `collectGarbage`), and since this doesn't capture any constructor-time state in the closure, it can be replaced with a simple function `private async initializeGCStateFromBaseSnapshot(currentReferenceTimestamp: number)`.  I made the change and stashed it, maybe I'll put a PR out after you merge.

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts (hunk 8 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts:405]: Yes, this can happen before trailing ops are processed for read connections. I am not planning to use the `IConnected` timestamp so that should be fine. However, this means until we get a new op and it connects / reconnects, we will keep using old timestamps.

I don't think there ia way to get this 100% correct with what we currently have. Once we have the "reference" op, we could drop `Timers` and do the unreferenced state calculation based on the timestamp of the op for which GC is called, that should solve most if not all of the issues. Even in case of loads, we could use the timestamp of the last processed op to know what the state of the node is. Haven't thought through that one fully.

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts (hunk 9 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts:408]: It's already logged for all events AFAIK.

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:1]: I think the fact that the change did not break anything is enough. I could add tests with trailing ops and ensure they are detected now as compared to before but I am not sure if its worth it.

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts (hunk 10 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts:177]: I was never a fan of validation via the `itExpects` block. IMO, if we want to validate these events, we should make it part of the test itself.

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts (hunk 11 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts:314]: The reason this statement existed was to explain why this was part of itExpects (not sure why only this was special though). It doesn't add any value anymore.

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts (hunk 10 above)
Comment (markfields) [reply packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts:177]: Yeah, I'll add a low-pri note to the follow-ups, to double-check if we have coverage using MockLogger anywhere and consider adding some.

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts (hunk 9 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts:409]: Looks to me like it's not, I sent you a direct message about it.

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts (hunk 8 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts:406]: Yeah need to give some more thought to how to drop the timers, that would be nice.  This approach makes sense for now, definite improvement and a simple model to reason over.

**Review by markfields — APPROVED** [review]

**Review by markfields — APPROVED** [review]

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts (hunk 4 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts:260]: Another reason to make this a lazy promise is to ensure it is re-entrant. Even if its called multiple times, the initialization should only happen once. There can be scenarios where a container keeps on reconnecting and this is called multiple times. Another scenario which is more concerning is when the summarizer client connects and starts GC immediately. There are probably other ways to make it eventual consistent, lazy promise is a simple one.
